src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1059 @@
+        assert_type(bsonVal, bsoncxx::type::k_array);
+        val = bsonVal.get_array();
+    }
+
+    /**
+     * Loads a BSON binary data value from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the binary data will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_binary& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_binary);
+        val = bsonVal.get_binary();
+    }
+
+    /**
+     * Loads a BSON ObjectID from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the OID will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_oid& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_oid);
+        val = bsonVal.get_oid();
+    }
+
+    /**
+     * Loads an ObjectID from the current node.
+     *
+     * @param val
+     *    The oid variable into which the OID will be loaded.
+     */
+    void loadValue(bsoncxx::oid& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_oid);
+        val = bsonVal.get_oid().value;
+    }
+
+    /**
+     * Loads a bool from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the bool will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_bool& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_bool);
+        val = bsonVal.get_bool();
+    }
+
+    /**
+     * Loads a BSON datetime from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the datetime will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_date& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_date);
+        val = bsonVal.get_date();
+    }
+
+    /**
+     * Loads a 32 bit int from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the int will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_int32& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_int32);
+        val = bsonVal.get_int32();
+    }
+
+    /**
+     * Loads a 64 bit int from the current node.
+     *
+     * @param val
+     *    The BSON variable into which the int will be loaded.
+     */
+    void loadValue(bsoncxx::types::b_int64& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_int64);
+        val = bsonVal.get_int64();
+    }
+
+    /**
+     * Loads a BSON datetime from the current node.
+     *
+     * @param val
+     *    The time_point variable into which the datetime will be loaded.
+     */
+    void loadValue(std::chrono::system_clock::time_point& val) {
+        auto bsonVal = search();
+        assert_type(bsonVal, bsoncxx::type::k_date);
+        val = std::chrono::system_clock::time_point(
+            std::chrono::duration_cast<std::chrono::system_clock::duration>(
```

Comment:
Is this duration_cast necessary? Can you just do 
`std::chrono::system_clock::time_point(std::chrono::milliseconds{bsonVal.get_date().value})`?